### ISSUE TITLE
READ BINARY command implementation

### ISSUE THREAD

**Issue by edbek** [tomasz-lisowski/swsim#4]
I started testing swsim, and this is what I found out when I tried to issue the READ BINARY command:

1) first, of course, I give the SELECT command:

> A0 A4 00 00 02 2F E2
< 9F 0F
> A0 C0 00 00 0F
< 00 00 00 0A 2F E2 04 00 00 00 00 01 02 00 00 90 00

swsim replies "9F0F" , which means that next you need to submit a GET RESPONSE. Next, we issue this command and everything is fine, that is, everything goes according to the GSM specification. OK.

2) then I give READ BINARY with the length of the file I need:

> A0 B0 00 00 0A
< 9F 0A        //  why is 9F 0A here, and not immediately an answer like 98 88 12 01 00 00 50 01 80 F4 90 00 ?    
> A0 C0 00 00 0A
< 98 88 12 01 00 00 50 01 80 F4 90 00

here the unexpected happens - swsim again issues "9FXX", which indicates that further it is necessary to submit GET RESPONSE again.
But I see in the GSM specification that READ BINARY must be answered immediately. 3GPP TS 11.11 version 8.14.0 has clause 9.4.7 and Table 11 which indicates that the READ BINARY command does not have a "9FXX" response.
But when I work with a real SIM card, then the answer from it is correct:

> A0 B0 00 00 0A
< 98 07 91 09 04 21 50 42 71 F1 90 00  // there is no 9F 0A here !

In addition, I saw two functions in your code that are responsible for reading binary files:
apduh_bin_read()
  and
apduh_gsm_bin_read()

and it looks like the last feature is working in the swsim project. 

Could you explain how to properly work with the READ BINARY command?
Thank you in advance !

**Comment by tomasz-lisowski** [comment]
I just took a look at GSM 11.11 v4.21.1 and indeed you are right. I think I overlooked this due to ESTI TS 102 221 v16.4.0 as that spec (for newer cards) allows for a 91XX status from READ BINARY, i.e., an indication equivalent to 9FXX on GSM. Interestingly, all the GSM devices I used to test swSIM accepted my incorrect implementation of READ BINARY, but I don't remember if other SIM cards did this as well. Anyway, thanks for pointing out yet another issue!

The upside of the current way APDU commands are handled is that to fix this, all you need to do is modify (inside `apduh_gsm_bin_read`):
```C
/* Enqueue the file data for later retrieval using GET RESPONSE. */
if (swicc_apdu_rc_enq(&swicc_state->apdu_rc, file->data,
              len_expected) != SWICC_RET_SUCCESS)
{
    res->sw1 = 0x92; /* "Memory problem." */
    res->sw2 = 0x40;
    res->data.len = 0U;
    return SWICC_RET_SUCCESS;
}

/* "Length 'XX' of the response data." where 'XX' is SW2. */
res->sw1 = 0x9F;
res->sw2 = len_expected;
res->data.len = 0U;
return SWICC_RET_SUCCESS;
```

To:

```C
/* Copy the file data to the response. */
memcpy(res->data.b, file->data, len_expected);

/* Send back the data. */
res->sw1 = 0x90;
res->sw2 = 0x00;
res->data.len = len_expected;
return SWICC_RET_SUCCESS;
```

To explain this a little further:
- Every `apduh` function handles a type of APDU command in some command class. This is why, as you pointed out, there are two READ BINARY functions. `apduh_bin_read` is for READ BINARY commands in the interindustry classes defined in ISO 7816, and `apduh_gsm_bin_read` is for the classes defined by 3GPP/ETSI for GSM.
- Every `apduh` function gets a pointer to the `res` (response) struct, which it must fill with its response to the command. You can see the regular SW1 and SW2 fields, but there is also a data field. The data field contains a buffer and a length, where the buffer obviously stores the data to return inside the RAPDU, and the length indicates how much data is in that buffer. For example, setting the data length field to 0 will send back just the status word (SW1:SW2) and that's what I did before; the fix just copies the response data into the response directly instead of first queuing it to an internal buffer used by GET RESPONSE.
- BTW, the memory copy in the new version is safe because the expected length and the actual file length are checked beforehand.

As a side note, setting the `sw1`, `sw2`, and `data.len` fields for the `res` struct is so common that I made a macro for it `SWICC_APDUH_RES`.

**Comment by edbek** [comment]
Thank you for the quick response !

Can you please tell me if there are such programs for simulating the operation of a GSM modem so that you can fully check swsim as if it were a real GSM modem?

**Comment by tomasz-lisowski** [comment]
The testing I've done was with real devices and using [FirmWire](https://github.com/FirmWire/FirmWire). 

To connect swSIM to a real device, you could get something like [SIMtrace2](https://shop.sysmocom.de/SIMtrace2-Hardware-Kit/simtrace2-kit), then install the [PC/SC driver for swICC](https://github.com/tomasz-lisowski/swicc-drv-ifd). 
PC/SC is the standardized interface for connecting smart cards to PCs and since SIM cards are a type of smart card, the driver will relay the swSIM messages through the PC/SC middleware and to the SIMtrace device which will forward the data (almost) verbatim to the physical device (and back). SIMtrace2 software is shipped with a tool that does this BTW (take a look [here](https://gitea.osmocom.org/sim-card/simtrace2/src/branch/master/host/src/simtrace2-cardem-pcsc.c)). 
TL;DR swSIM <-> PC/SC driver for swICC <-> PC/SC middleware <-> SIMtrace2 cardem <-> Physical device.

The alternative is FirmWire which allows you to emulate the modem in-software. The downside is of course that you'd need to reverse engineer the SIM card interface of the modem you want to test with and implement a driver for it. This is a matter of using the tools shipped with FirmWire to create a new peripheral that handles memory reads/writes to certain memory locations (the hardware peripherals are just memory-mapped to certain portions of memory). For example, [here](https://github.com/FirmWire/FirmWire/blob/main/firmwire/vendor/shannon/hw/uart.py) is a UART peripheral included with FirmWire for the Shannon basebands.
